add set functions

cseed opened the pull request:
Added the straightforward implementations to get them in the IRFunctionRegistry.  They are not optimal from the computational complexity perspective (e.g. union is n log m instead of n + m).  I will improve them in a separate PR.

maccum commented on src/test/scala/is/hail/TestUtils.scala:
```diff
@@ -249,8 +250,65 @@ object TestUtils {
+    val argsType = TTuple(args.map(_._2): _*)
+    val resultType = TTuple(x.typ)
+
+    def rewrite(x: IR): IR = {
```
`rewrite()` isn't being used. delete?

cseed commented:
You're the reviewer: you're the boss.  I switched to the second style.  I also added missingness cases to the Scala tests, found and fixed a couple of bugs, including the FilterArray behavior (if the condition is true, even on NA elements, the NA elements should remain).  Added corresponding tests.
